src/app-layer-ftp.c

```diff
@@ -1439,36 +1439,37 @@ uint16_t JsonGetNextLineFromBuffer(const char *buffer, const uint16_t len)
     return c == NULL ? len : c - buffer + 1;
 }
 
-json_t *JsonFTPDataAddMetadata(const Flow *f)
+void JsonFTPDataAddMetadata(const Flow *f, JsonBuilder *jb)
 {
     const FtpDataState *ftp_state = NULL;
     if (f->alstate == NULL)
-        return NULL;
+        return;
+
     ftp_state = (FtpDataState *)f->alstate;
-    json_t *ftpd = json_object();
-    if (ftpd == NULL)
-        return NULL;
+
+    jb_open_object(jb, "ftp-data");
     if (ftp_state->file_name) {
         size_t size = ftp_state->file_len * 2 + 1;
-        char string[size];
-        BytesToStringBuffer(ftp_state->file_name, ftp_state->file_len, string, size);
-        json_object_set_new(ftpd, "filename", SCJsonString(string));
+        char fname[size];
+        BytesToStringBuffer(ftp_state->file_name, ftp_state->file_len, fname, size);
+        jb_set_string(jb, "filename", fname);
     }
     switch (ftp_state->command) {
         case FTP_COMMAND_STOR:
-            json_object_set_new(ftpd, "command", json_string("STOR"));
+            jb_set_string(jb, "command", "STOR");
             break;
         case FTP_COMMAND_RETR:
-            json_object_set_new(ftpd, "command", json_string("RETR"));
+            jb_set_string(jb, "command", "RETR");
```

Comment:
@jasonish I still think we can optimize these constant cases with some preproc magic to become a single string that we append

Reply:
Yeah, need to look at this again.